### PR TITLE
prevent DAG callback exception from crashing scheduler

### PR DESCRIPTION
**Pull request by houqp**
Also adds type annotation so we can catch signal type exceptions at build time.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Review by potiuk — CHANGES_REQUESTED** [review]
Nice! But I think this is sort-of breaking change so we should likely make a note about it in the UPDATING.md

**Review by turbaszek — APPROVED** [review]
Looks good but as Jarek mentioned we should add a not to UPDATING.md. I like the idea of Context type!

**Comment by houqp** [comment]
Added entry in UPDATING.md as well as added a counter metric to help better monitor this issue.

**Review by turbaszek — APPROVED** [review]

**Comment by houqp** [comment]
@turbaszek @potiuk fixed all test errors, ready for another round of review.

**Review by turbaszek — APPROVED** [review]

File: airflow/models/dag.py [hunk 1]
```diff
@@ -686,7 +689,11 @@ def handle_callback(self, dagrun, success=True, reason=None, session=None):
             ti.task = self.get_task(ti.task_id)
             context = ti.get_template_context(session=session)
             context.update({'reason': reason})
-            callback(context)
+            try:
+                callback(context)
+            except Exception:
+                self.log.exception("failed to invoke dag state update callback")
+                Stats.incr("dag.callback_exceptions")
```
Comment (mik-laj) [inline airflow/models/dag.py:696]: Can you add this key to docs? 
https://airflow.readthedocs.io/en/latest/metrics.html

Comment (houqp) [reply airflow/models/dag.py:696]: good call, i always forgot about that doc page :)

**Review by potiuk — APPROVED** [review]

**Pull request merged**